src/libs/antares/study/parts/hydro/finallevelinflowsmodifyer.h

```diff
@@ -0,0 +1,90 @@
+/*
+** Copyright 2007-2023 RTE
+** Authors: Antares_Simulator Team
+**
+** This file is part of Antares_Simulator.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** There are special exceptions to the terms and conditions of the
+** license as they are applied to this software. View the full text of
+** the exceptions in file COPYING.txt in the directory of this software
+** distribution
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with Antares_Simulator. If not, see <http://www.gnu.org/licenses/>.
+**
+** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
+*/
+#ifndef __ANTARES_LIBS_STUDY_PARTS_HYDRO_FINAL_LEVEL_INFLOWS_MODIFYER_H__
+#define __ANTARES_LIBS_STUDY_PARTS_HYDRO_FINAL_LEVEL_INFLOWS_MODIFYER_H__
+
+#include <vector>
+#include <yuni/yuni.h>
+#include <antares/study/parameters.h>
+#include <antares/array/matrix.h>
+
+namespace Antares::Data
+{
+class PartHydro;
+/*!
+ ** \brief Final Reservoir Level data for a single area
+ */
+class FinalLevelInflowsModifier
+{
+public:
+    FinalLevelInflowsModifier() = delete;
+    FinalLevelInflowsModifier(const PartHydro& hydro,
+                              const unsigned int& areaIndex,
+                              const AreaName& areaName);
+    void initialize(const Matrix<double>& scenarioInitialHydroLevels,
+                    const Matrix<double>& scenarioFinalHydroLevels,
+                    const uint lastSimulationDay,
+                    const uint nbYears);
+
+    bool CheckInfeasibility(uint year);
+    bool isApplicable(uint year);
+
+    // vectors containing data necessary for final reservoir level calculation
+    // for one area and all MC years
+    // vector indexes correspond to the MC years
+    std::vector<double> deltaLevel;
+
+
+private:
```

Comment:
**Idea for a possible refactoring** 
I really hesitated to do it, but eventually resisted, partially for lack of time, and because some questions remain.
Class **FinalLevelInflowsModifier** does checks and allows computing a final level for a given area.
But we can see that a lot of public and private methods have the year as argument.
I see it as a sign that this class should either : 
- become mainly a container for a list  of yearly final level inflow modifier (let's call it class **YearlyLvlModfier** for instance).
- replaced by a list of **YearlyLvlModfier**.

Maybe there would a benefit to keep **FinalLevelInflowsModifier** as a container, for example that we would not have to change the way it's used in the production code or unit tests. Rather we would change the code of **FinalLevelInflowsModifier** internally.

Anyway, the purpose if this refactoring would be to have classes with less responsibilities, and possibly only one. As usual, it would make code : 
- more clear (it maybe would introduce more classes, but smaller and so easier to understand)
- more changeable, as small classes usually facilitates this
- more testable (we could test **YearlyLvlModfier** in isolation, and, as it is supposed to be smaller than **FinalLevelInflowsModifier**, would be unit tested more easily). Same thing for **FinalLevelInflowsModifier**, possibly much smaller.

Reply:
I have to read this more carefully tomorrow. If we keep the class just as the container or switch to a list. 
Where do we move out all FinalLevelInflowsModifier - methods?